Prevent ALTERing non-existing KS with tablets

ptrsmrn opened the pull request:
ALTER tablets KS executes in 2 steps:
1. ALTER KS's cql handler forms a global topo req, and saves data required to execute this req,
2. global topo req is executed by topo coordinator, which reads data attached to the req.

The KS name is among the data attached to the req. There's a time window between these steps where a to-be-altered KS could have been DROPped, which results in topo coordinator forever trying to ALTER a non-existing KS. In order to avoid it, the code has been changed to first check if a to-be-altered KS exists, and if it's not the case, it doesn't perform any schema/tablets mutations, but just removes the global topo req from the coordinator's queue.
BTW. just adding this extra check resulted in broader than expected changes, which is due to the fact that the code is written badly and needs to be refactored - an effort that's already planned under #19126
(I suggest to disable displaying whitespace differences when reviewing this PR).

Fixes: #19576

Requires 6.0 backport

ptrsmrn commented:
@kbr-scylla I was thinking about adding a UT for this change, something along the lines of:
```
# test/cql-pytest/test_keyspace.py
def test_alter_dropped_tablets_keyspace(cql, this_dc, skip_without_tablets):
    keyspace = "to_be_altered_keyspace"
    cql.execute(f"CREATE KEYSPACE {keyspace} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}' : 1 }} AND TABLETS = {{'enabled': true}}")
    cql.execute_async(f"ALTER KEYSPACE {keyspace} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', '{this_dc}': 1}}")
    cql.execute(f'DROP KEYSPACE IF EXISTS {keyspace}')
```
which, theoretically, would create a time window allowing to drop a KS that hasn't been ALTERed by the topo coordinator yet, which is achieved by scheduling ALTER call asynchronously (not waiting for it to complete), so that the following DROP statement could interfere with the above's ALTER. But executing this test 10k times didn't reproduce the issue seen in #19576, so it's not that easy to reproduce it + I think that we already have a testcase that is able to test this scenario - exactly the same test that discovered it (#19576).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10299/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10299/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2525/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10550/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10299/)
#### Build Details:
- Duration: 2 hr 47 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
@ptrsmrn relying on longevity tests for regression testing is not a good idea.

The PR should come with a regression test -- best would be to have one that fails before the fix, and does not fail after.

You should be able to **very easily** orchestrate the scenario of:
- start ALTER, it commits the request for coordinator to pick up
- perform DROP
- coordinator picks up the request

Using our error injection framework. I don't know if this framework is accessible from cql-pytest, if not, just write a topology_custom test (even single node test should suffice IIUC).

You can inject a wait into topology coordinator at the beginning of handling the request. This injection will wait for a message from the test. Then the test will perform a DROP. Then it will message the injection so coordinator unblocks and handles the request.
See:
-  `future<> inject(const std::string_view& name, waiting_handler_fun func, bool share_messages)` in utils/error_injection.hh
- test/topology_custom/test_raft_no_quorum.py: `test_quorum_lost_during_node_join` for example usage. In particular how `injection_handler` is used

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -773,35 +771,41 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
-                        updates.clear(); // remove all tablets mutations ...
-                        break;           // ... and only create mutations deleting the global req
-                    }
+                    for (const auto& table : ks.metadata()->tables()) {
```
Please split indentation changes to a separate commit so the PR is easier to review.

ptrsmrn replied:
I left a note in the description that touches on whitespaces subject:
> (I suggest to disable displaying whitespace differences when reviewing this PR).

So now, so that I remember it once and for all, are maintainers, you included, good with just disable whitespaces diff or is that I always should extract the "refactor" part into a separate commit?

kbr-scylla replied:
Interesting, I didn't know about this feature.

I don't know what to decide and don't have a strong opinion. One potential reason for extracting whitespace changes to separate commits is that it improves experience when browsing the `git log` -- which I sometimes do when e.g. debugging a regression, and trying to find where it happened. In theory I could in that case find the PR and there use the whitespace hiding feature, but that's more jumping and I don't know.

@scylladb/scylla-maint any opinions on this matter?

ptrsmrn replied:
I'm not sure about `git log`, but `git blame` has extra options to ignore whitespace changes and detect movements of code within/outside of a file, so refactoring the code, i.e. modifying whitespaces or moving it, shouldn't make it that harder to track down a bug.

denesb replied:
But I don't want to disable whitespace changes when reviewing PRs, making sure whitespace is just right is the single most important duty of a maintainer!
On a serious note, I don't think ignoring whitespace while reviewing is a good idea. It is a good idea when using git blame.
So let's keep splitting reformatting into separate patches, as before.

xemul replied:
I second Botond's suggestion -- though toggling display of whitespaces differences is possible, it doesn't make review easier

ptrsmrn replied:
Thanks, I split the changes.

ptrsmrn commented:
@kbr-scylla thanks for the feedback, suggestions applied, please re-review.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10689/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10689/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10689/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2855/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10877/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10689/)
#### Build Details:
- Duration: 3 hr 56 min
- Builder: spider5.cloudius-systems.com

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -803,6 +804,9 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
+                    error = "Can't ALTER keyspace " + ks_name + ", keyspace doesn't exist";
+                }
 
                 updates.push_back(canonical_mutation(topology_mutation_builder(guard.write_timestamp())
```
Shouldn't we skip entering `tablet_migration` and increasing topology version if ALTER fails (there's an `error`)?

kbr-scylla replied:
I see that this problem is pre-existing, but I'd like to understand why is this done so nevertheless, and maybe it needs a separate issue.

ptrsmrn replied:
If preparing ALTER mutations (schema mutations or tablets mutations) fails, we still need to delete the global `rf_change` req along with the topo req tracking entry. 
So whether ALTER succeeds or not, we need to perform some mutations.

kbr-scylla replied:
This doesn't answer my question. Yes we need to perform **some** mutations, the ones you mentioned. But what about these things that I asked:
> entering tablet_migration and increasing topology version

?

ptrsmrn replied:
That needs to be investigated, I don't know the answer off the top of my head. Added to https://github.com/scylladb/scylladb/issues/19123

kbr-scylla commented on service/topology_coordinator.cc:
```diff
@@ -2898,6 +2898,12 @@ future<> topology_coordinator::run() {
                 co_await await_event();
                 rtlogger.debug("topology coordinator fiber got an event");
             }
+            co_await utils::get_local_injector().inject("wait-before-handling-keyspace-rf-change", [] (auto& handler) -> future<> {
```
Let's name this injection more generically. It has nothing to do with keyspace RF change -- it's a general injection to sleep after getting event in topology coordinator fiber. Currently it's only used in test for keyspace rf change, but we may want to reuse it in other tests in other contexts in the future.

ptrsmrn replied:
Right, changed to `wait-after-topology-coordinator-gets-event`.

kbr-scylla commented on test/topology_custom/test_tablets.py:
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
+
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_alter_dropped_tablets_keyspace(manager: ManagerClient) -> None:
```
Let's put this in separate file so the test case can run concurrently with other test cases in test_tablets.py.

Also, does this test fail without the fix?

ptrsmrn replied:
Yes, the test fails without the fix.

ptrsmrn replied:
What's the reason for moving this specific testcase to a separate file? You want to test if ALTER can indeed run concurrently with other operations? Testing from this angle sounds OK, but this approach may not be the best, understanding why a testcase failed would be hard.

kbr-scylla replied:
> What's the reason for moving this specific testcase to a separate file?

Speed of the test suite i.e. how fast test suite can finish (the more test cases can run concurrently, the faster the whole suite will finish)

> You want to test if ALTER can indeed run concurrently with other operations?

No, each test case has exclusive access to a leased cluster...

ptrsmrn replied:
Clear, thanks, done.

kbr-scylla commented on test/topology_custom/test_tablets.py:
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
+
+    async def alter_tablets_ks_without_waiting_to_complete():
+        res = await manager.get_cql().run_async("select data_center from system.local")
+        this_dc = res[0].data_center  # ALTER tablets KS needs to have the DC specified explicitly
```
Why do you need to get DC from CQL though? Note that the node you get DC from, and the one you execute the statement later, could be different, so if they booted in different DCs, this code would not make much sense. Well it's not a problem in this test because they should boot in the same DC. I guess the only thing you want to do here is to ensure you pass a DC that exists? (and don't care which DC in particular it is) You could make it more obvious in the comments.

ptrsmrn replied:
I need to explicitly state the name of the DC, because ALTER tablets KS doesn't accept the generic `replication_factor` tag.

ptrsmrn replied:
Changed the comment to be more specific: `# ALTER tablets KS only accepts a specific DC, it rejects the generic 'replication_factor' tag`

kbr-scylla commented on test/topology_custom/test_tablets.py:
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
+        res = await manager.get_cql().run_async("select data_center from system.local")
+        this_dc = res[0].data_center  # ALTER tablets KS needs to have the DC specified explicitly
+        manager.get_cql().run_async("alter keyspace ks "
+                                    f"with replication = {{'class': 'NetworkTopologyStrategy', '{this_dc}': 1}}")
```
We should `await` for it and verify that it returns an error result at the end of the test.

You decided to spawn a task (`asyncio.create_task`) so the test case an progress concurrently while this is happening.

At the end of the test we can `await` the returned task:
```
task = asyncio.create_task(alter_tablets...)

... do the other stuff, drop keyspace, message handler ...

try:
    await task
except:
   ... verify that the returned exception is as expected (failed ALTER) ...
else:
   fail the test: we expected an exception, not successful response
```

ptrsmrn replied:
Rewritten, thanks.

kbr-scylla commented on test/topology_custom/test_tablets.py:
```diff
@@ -301,3 +301,51 @@ async def test_read_of_pending_replica_during_migration(manager: ManagerClient,
+                                    f"with replication = {{'class': 'NetworkTopologyStrategy', '{this_dc}': 1}}")
+
+    # by creating a task this way we ensure it's immediately executed, but we won't wait until it's completed
+    await asyncio.create_task(alter_tablets_ks_without_waiting_to_complete())
```
No point to `create_task` and then await it immediately in the foreground... This is equivalent to `await alter_tablets_ks_without_waiting_to_complete()`. But see my comments above. We should have a task but keep it in background and await at the end to verify the result.

ptrsmrn replied:
Addressed in https://github.com/scylladb/scylladb/pull/19666#discussion_r1707772916

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10817/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10817/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10817/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10817/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10976/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2953/)
#### Build Details:
- Duration: 4 hr 4 min
- Builder: spider3.cloudius-systems.com

ptrsmrn commented:
@kbr-scylla can you please re-eview?

kbr-scylla commented:
LGTM except:
- https://github.com/scylladb/scylladb/pull/19666#discussion_r1715208314
- https://github.com/scylladb/scylladb/pull/19666#discussion_r1715206430 (pre-existing problem, but I think an issue needs to be created)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10904/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10904/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_tablets_cql
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10904/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3036/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10904/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11059/)
#### Build Details:
- Duration: 2 hr 15 min
- Builder: spider4.cloudius-systems.com

ptrsmrn commented:
> LGTM except:
> 
> * [Prevent ALTERing non-existing KS with tablets #19666 (comment)](https://github.com/scylladb/scylladb/pull/19666#discussion_r1715208314)
> * [Prevent ALTERing non-existing KS with tablets #19666 (comment)](https://github.com/scylladb/scylladb/pull/19666#discussion_r1715206430) (pre-existing problem, but I think an issue needs to be created)

@kbr-scylla addressed the comments, the CI is green.
